builtin/logical/pki/secret_certs.go

```diff
@@ -0,0 +1,59 @@
+package pki
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+// SecretCertsType is the name used to identify this type
+const SecretCertsType = "pki"
+
+func secretCerts(b *backend) *framework.Secret {
+	return &framework.Secret{
+		Type: SecretCertsType,
+		Fields: map[string]*framework.FieldSchema{
+			"certificate": &framework.FieldSchema{
+				Type: framework.TypeString,
+				Description: `The PEM-encoded concatenated certificate and
+issuing certificate authority`,
+			},
+			"private_key": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "The PEM-encoded private key for the certificate",
+			},
+			"serial": &framework.FieldSchema{
+				Type: framework.TypeString,
+				Description: `The serial number of the certificate, for handy
+reference`,
+			},
+		},
+
+		DefaultDuration:    168 * time.Hour,
+		DefaultGracePeriod: 10 * time.Minute,
+
+		Revoke: b.secretCredsRevoke,
+	}
+}
+
+func (b *backend) secretCredsRevoke(
```

Comment:
If the certificate is revoked just due to age (e.g. expiration time hit) then we probably do not need to generate a CRL entry. The underlying certificate is already invalid anyways. It's only for an early revocation that CRL must be generated. (Correct me if I'm wrong)